### PR TITLE
call functions and get responses

### PR DESCRIPTION
**Pull request by tobowers**
This pull request adds a relatively simple way to expose functions to the corestep. You can force the corestep to call functions or use the normal openAI "auto" feature - as seen in the example here.

I don't suspect this will be the ultimate API, but at least it unlocks toying with functions in a way that has a consistent API.

### CODE REVIEW

**Comment by tobowers** [comment]
@kafischer

File: core/src/cortexStep.ts [hunk 1]
```diff
@@ -194,6 +217,38 @@ export class CortexStep {
...
+    if (fn === undefined) {
+      console.error("functions: ", this.functions, "does not include ", functionCall)
+      throw new Error(`function ${functionCall.name} not found`);
+    }
+
+    const memories = await fn.run(JSON.parse(functionCall.arguments || "undefined")) || []
+
+    return new CortexStep(this.entityName, {
+      pastCortexStep: this,
+      lastValue: memories.slice(-1)[0].content,
```
Comment (kafischer) [inline core/src/cortexStep.ts:248]: The function runner should probably return the lastValues well as the memories

File: core/examples/example5_functions.ts [hunk 2]
```diff
@@ -0,0 +1,85 @@
+import * as readline from "readline";
+import { Blueprints, ChatMessage } from "../src";
+import dotenv from "dotenv";
+import { Action, CortexStep } from "../src/cortexStep";
+import { ChatMessageRoleEnum } from "../src";
+import { FunctionRunner } from "../src/languageModels/functions";
+
+dotenv.config();
+
+// this example demonstrates adding function calls to a CoreStep.
```
Comment (kafischer) [inline core/examples/example5_functions.ts:10]: This example isn't really interactive, should probably just be an integration test?

**Review by kafischer — CHANGES_REQUESTED** [review]
Very cool!

I'm not sure I understand the api being withFunctions versus requiring the FunctionRunner to always be specified in a .next step- it feels like it breaks something conceptually to me

**Comment by kafischer** [comment]
Or at least like a .nextWithFuctions so the effect is isolated?

**Comment by tobowers** [comment]
> Very cool!
> 
> I'm not sure I understand the api being withFunctions versus requiring the FunctionRunner to always be specified in a .next step- it feels like it breaks something conceptually to me

I can see that. OpenAI seems to think functions live above the conversation instead of part of the conversation, but I can see how that's a little different than here. I think conceptually to me it makes sense that they are similar to memories, but I adjusted to make it more incremental and part of the `next()` call now.

**Comment by tobowers** [comment]
> Or at least like a .nextWithFuctions so the effect is isolated?

Made this an optional addition to the .next() call

**Comment by kafischer** [comment]
I'm sure you're seeing something here that I'm not quite yet - I am interested to understand your point

**Review by kafischer — APPROVED** [review]

**Pull request merged**